File: src/ros2_python_tutorial/ros2_python_tutorial/p1_publisher.py
import rclpy
from rclpy.node import Node

from std_msgs.msg import String


class MinimalPublisher(Node):

    def __init__(self):
        super().__init__('minimal_publisher')
        self.fast_publisher_ = self.create_publisher(String, 'fast_topic', 10)
        self.slow_publisher_ = self.create_publisher(String, 'slow_topic', 10)
        fast_timer_period = 0.5  # seconds
        slow_timer_period = 1.0  # seconds
        self.fast_timer_ = self.create_timer(fast_timer_period, self.fast_timer_callback)
        self.slow_timer_ = self.create_timer(slow_timer_period, self.slow_timer_callback)
        self.fast_count_ = 0
        self.slow_count_ = 0

    def fast_timer_callback(self):
        msg = String()
        msg.data = 'Fast timer %d' % self.fast_count_
        self.fast_publisher_.publish(msg)
        self.get_logger().info('[Fast] Publishing: "%s"' % msg.data)
        self.fast_count_ += 1

    def slow_timer_callback(self):
        msg = String()
        msg.data = 'Slow timer %d' % self.slow_count_
        self.slow_publisher_.publish(msg)
        self.get_logger().info('[Slow] Publishing: "%s"' % msg.data)
        self.slow_count_ += 1

def main(args=None):
    rclpy.init(args=args)

    minimal_publisher = MinimalPublisher()

    rclpy.spin(minimal_publisher)

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    minimal_publisher.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/ros2_cpp_tutorial/src/p1_publisher.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

class MinimalPublisher : public rclcpp::Node
{
    public:
        MinimalPublisher()
        : Node("minimal_publisher"), fast_count_(0), slow_count_(0)
        {
            fast_publisher_ = this->create_publisher<std_msgs::msg::String>("fast_topic", 10);
            slow_publisher_ = this->create_publisher<std_msgs::msg::String>("slow_topic", 10);
            fast_timer_ = this->create_wall_timer(
                500ms, std::bind(&MinimalPublisher::fast_timer_callback, this));
            slow_timer_ = this->create_wall_timer(
                1s, std::bind(&MinimalPublisher::slow_timer_callback, this));
        }
    
    private:
        void fast_timer_callback()
        {
            auto message = std_msgs::msg::String();
            message.data = "Faster timer " + std::to_string(fast_count_++);
            RCLCPP_INFO(this->get_logger(), "[Fast] Publishing: '%s'",message.data.c_str());
            fast_publisher_->publish(message);
        }
        void slow_timer_callback()
        {
            auto message = std_msgs::msg::String();
            message.data = "Slow timer " + std::to_string(slow_count_++);
            RCLCPP_INFO(this->get_logger(), "[Slow] Publishing: '%s'", message.data.c_str());
            slow_publisher_->publish(message);
        }
        rclcpp::TimerBase::SharedPtr fast_timer_, slow_timer_;
        rclcpp::Publisher<std_msgs::msg::String>::SharedPtr fast_publisher_, slow_publisher_;
        size_t fast_count_, slow_count_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<MinimalPublisher>());
    rclcpp::shutdown();
    return 0;
}